### ISSUE TITLE
Database corrupt after listings update

### ISSUE THREAD

**Issue by stevehayles** [jellyfin/jellyfin#1952]
**Describe the bug**
After refreshing the guide data (from Schedules Direct or an XMLTV source) the database ends up with all entities of type LiveTvProgram, LiveTvChannel or UserView having a TopParentId field set to NULL.  

This effectively removes all LiveTv from the user interface and you can no longer access any channels, look at the guide or any related functionality

**To Reproduce**
1. Setup an HdHomeRun tuner with the UK schedules direct source set to Freeview
2. Let the initial installation complete the channel updates and refresh the EPG
3. At this stage everything will work as intended
4. Manually update the giude data from the dashboard under LiveTv

**Expected behavior**

The guide data should update in the background and leave the system in a usable state. 

**System**
 - OS: Tested on Windows 10 and Debian Buster via a Linux Docker host
 - Browser: Not applicable, but front end symptoms will show in any browser or client
 - Jellyfin Version: Compiled source - Master Branch
 - Reverse proxy: No Proxy

**Additional context**
The user interface issues are probably by design as the call on the backend gets to 'GetRecommendedProgramsInternal' in the LiveTvManager which includes a TopParentId field in the query.  This returns nothing because of the corrupted state of the database.

The initial guide update on a fresh install is always OK and not every update results in corruption.  Writing the correct TopParentId (that matches the Id of the LiveTv folder) into every LiveTvProgram or LiveTvChannel row, using a manual SQL update, gets the system back to a usable condition

**Comment by anthonylavado** [comment]
Hrm. I’m using 10.4.0 release with a TVHeadend>Antennas (HDHomeRun emulator) setup, using SchedulesDirect (in Canada), and it seems to be fine. So it must be something in the master branch then.

**Comment by stevehayles** [comment]
@anthonylavado Thanks for the update, it's a very difficult one to track down but I am making progress.  It definitely doesn't happen every update, which makes it hard but for sure I am seeing some logic issues which are related.

During a guide refresh I have seen that in the SaveItems call of SqliteItemRepository around line 589 there is a call `var topParent = item.GetTopParent();` which should definitely return the 'livetv' folder but actually returns null.

The reason for this from what I can see is that as the folder hierachy is recursed to get the parent everything works OK but the 'IsTopParent' property is checked and returns false for livetv when it should return true.

In turn this is related to the CollectionType not being set properly on the folder.

**Comment by anthonylavado** [comment]
Interesting! Can anyone else from @jellyfin/backend comment?

**Comment by JustAMan** [comment]
@stevehayles good find, awaiting PR! And I've edited your comment to enable C# syntax highlighting, so it's nicer to read.

**Comment by stevehayles** [comment]
@JustAMan Thanks for the code highlighting tip, I could not remember how to set it. Pull release is ready

**Comment by Bond-009** [comment]
@stevehayles Could you please try my fix and see if it works?
https://github.com/jellyfin/jellyfin/pull/1965

**Comment by stevehayles** [comment]
@Bond-009 I have just checked quickly with only the following line changed as the rest of the changes looked unrelated

```c#
saveItemStatement.TryBind("@data", JsonSerializer.SerializeToUtf8Bytes(item, type, _jsonOptions));
```
I wiped out my library.db to start again.  Seems perfect, the 'data' field is now populated with Json that contains the ViewType which in turn fixes the problem I was seeing.  I will apply the full change to double check.

Only issue I see is for those who already have a populated database which doesn't contain the ViewType correctly serialised but I guess they can re-add their Tuner and Listings provider.  Feels like the proper fix though. Nice!
